### DISCUSSION TITLE
[Broken] CROSS_PROCESS in istio environment.

### DISCUSSION THREAD

**Discussion by StanleyWang-GitHub** [apache/skywalking#8717] [Q&A]
Cluster Info:
skywlaking-oap: v8.9.1 installed by helm.
istio: v1.12.8 installed by istioctl.
openresty: v1.19.9 with skywalking-nginx-lua agent: v0.6.0

Problem:
I got Broken spans in tracing model.
![image](https://user-images.githubusercontent.com/13164735/159208629-43df8e49-7439-4358-81f2-62044ec38929.png)
![image](https://user-images.githubusercontent.com/13164735/159209195-f3285462-47d0-4ea8-96c8-e0369558ff85.png)

Istio MeshConfig pieces:
![image](https://user-images.githubusercontent.com/13164735/159208885-1c73a889-53cc-422e-8db2-708d44543e9f.png)

Openresty nginx.conf pieces:
```
    lua_package_path "/usr/local/openresty/lua/lib/?.lua;;";

    # Buffer represents the register inform and the queue of the finished segment
    lua_shared_dict tracing_buffer 100m;
    
    # Setup coredns to resolve oap-server service name.
    # Init is the timer setter and keeper
    # Setup an infinite loop timer to do register and trace report.
    resolver 192.168.0.10 valid=5 ipv6=off;
    init_worker_by_lua_block {
        local metadata_buffer = ngx.shared.tracing_buffer

        -- Set service name
        metadata_buffer:set('serviceName', 'frontend')
        -- Instance means the number of Nginx deployment, does not mean the worker instances
        metadata_buffer:set('serviceInstanceName', 'frontend-instance-01')
        -- type 'boolean', mark the entrySpan include host/domain
        metadata_buffer:set('includeHostInEntrySpan', false)

        -- set random seed
        require("skywalking.util").set_randomseed()
        require("skywalking.client"):startBackendTimer("skywalking-opa.skywalking.svc.cluster.local")

        -- If there is a bug of this `tablepool` implementation, we can
        -- disable it in this way
        -- require("skywalking.util").disable_tablepool()

       skywalking_tracer = require("skywalking.tracer")
    }

    server {
        listen 80 default_server;
        listen [::]:80 ipv6only=on;
        server_name _;
        access_log /var/log/nginx/access.log json_combined;
        access_log /usr/local/openresty/nginx/logs/access.log json_combined;

        location / {
            default_type text/html;
            root /usr/local/openresty/nginx/html;
            index index.html index.htm;

            rewrite_by_lua_block {
                skywalking_tracer:start("frontend service");
            }

            body_filter_by_lua_block {
                if ngx.arg[2] then
                    skywalking_tracer:finish()
                end
            }

            log_by_lua_block {
                skywalking_tracer:prepareForReport()
            }
        }
        
        location /metrics {
            vhost_traffic_status_display;
            vhost_traffic_status_display_format prometheus;
            access_log off;
        }
        
        location = /status
        {
            allow 127.0.0.1/32;
            deny all;
            stub_status on;
            access_log off;
        }

        error_page 500 501 502 503 504 /50x.html;
        location = /50x.html {
          root /usr/local/openresty/nginx/html;
        }
    }
```

**Comment by wu-sheng** [comment]
We never used OpenResty with Istio, you may try to ask at the APISIX community? They are integrating APISIX(on the top of OpenResty) with Istio/Envoy together.

**Reply by WilliamShoww** [reply]
> We never used OpenResty with Istio, you may try to ask at the APISIX community? They are integrating APISIX(on the top of OpenResty) with Istio/Envoy together.

那我使用[skywalking-nginx-lua](https://github.com/apache/skywalking-nginx-lua)的到tracer数据也是这样的，Nginx和后面的服务串不起来，怎么解决？各处也没找到官方文档

**Reply by wu-sheng** [reply]
@WilliamShoww All things on GitHub must be in English. The doc is on GitHub, how to resolve your local issue, we don't know if you didn't check what is going on.
FAQ about `boken` VNode is very clear. https://skywalking.apache.org/docs/main/v8.9.1/en/faq/vnode/

**Reply by WilliamShoww** [reply]
> @WilliamShoww All things on GitHub must be in English. The doc is on GitHub, how to resolve your local issue, we don't know if you didn't check what is going on. FAQ about `boken` VNode is very clear. https://skywalking.apache.org/docs/main/v8.9.1/en/faq/vnode/

I want to know how to use options(skywalking_tracer:start("upstream service", {custom = "custom_value"})) in [skywalking-nginx-lua](https://github.com/apache/skywalking-nginx-lua)? could you tell me where can be found doc about it.

**Reply by wu-sheng** [reply]
This option has nothing related to whether the trace is continuous or not.

> skywalking_tracer:start("upstream service", {custom = "custom_value"})

This is already how to use, what document do you expect?

**Reply by WilliamShoww** [reply]
> This option has nothing related to whether the trace is continuous or not.
> 
> > skywalking_tracer:start("upstream service", {custom = "custom_value"})
> 
> This is already how to use, what document do you expect?

Sorry, I thought setting this option can make the tracer continuous. It seems that my opinion is wrong. Nginx is special, it's tracer can't  continuous with downstrem server.

**Reply by wu-sheng** [reply]
The repo you are looking at, includes e2e, main repo has e2e too. Both of them have verified trace and topology works, as well as apisix project checked too.

**Reply by wu-sheng** [reply]
Unless you could find something locally, it is impossible for me to guess.

**Reply by WilliamShoww** [reply]
> Unless you could find something locally, it is impossible for me to guess.

I found why the tracer is not continuous，because i setting downstream java server agent's agent_name_space, so downstream agent get upstream server span info from headers key, the headers key is not default sw8, finally key is AGENT_NAMESPACE + "-sw8". 
Now after i change the [skywalking-nginx-lua](https://github.com/apache/skywalking-nginx-lua)'s code, tracer is continuous.
So i suggest the [skywalking-nginx-lua](https://github.com/apache/skywalking-nginx-lua) also provide AGENT_NAMESPACE setting options.

**Reply by wu-sheng** [reply]
`AGENT_NAMESPACE` is removed from Java agent already. It wouldn't work like this since 8.10.0

**Comment by StanleyWang-GitHub** [comment]
Is there anyone tested if nginx behind an envoy proxy? The problem (Broken VNode)I met was in istio data plane circumstance. I think the 2 VNODE shown in UI should be envoy ingress and egress listener.
